scalding-core/src/main/scala/com/twitter/scalding/estimation/memory/MemoryHistoryEstimator.scala

```diff
@@ -0,0 +1,99 @@
+package com.twitter.scalding.estimation.memory
+
+import com.twitter.scalding.estimation.{ FlowStepHistory, FlowStrategyInfo, HistoryEstimator, MemoryEstimatorConfig, Task }
+import org.apache.hadoop.mapred.JobConf
+import org.slf4j.LoggerFactory
+
+// Tuple(MapMemory in MB, ReduceMemory in MB), or None to keep the default.
+case class MemoryEstimate(mapMemory: Option[Long], reduceMemory: Option[Long])
+
+object MemoryHistoryEstimator {
+  val CommittedHeapBytes = "COMMITTED_HEAP_BYTES"
+  val CpuMs = "CPU_MILLISECONDS"
+  val PhysicalMemoryBytes = "PHYSICAL_MEMORY_BYTES"
+  val GCTimeMs = "GC_TIME_MILLIS"
+
+  implicit class MemoryRichTask(val task: Task) extends AnyVal {
+    def committedHeapBytes: Option[Long] = task.counters.get(CommittedHeapBytes)
+  }
+}
+
+trait MemoryHistoryEstimator extends HistoryEstimator[MemoryEstimate] {
+  import MemoryHistoryEstimator.MemoryRichTask
+
+  private val LOG = LoggerFactory.getLogger(this.getClass)
+
+  override def maxHistory(conf: JobConf): Int = MemoryEstimatorConfig.getMaxHistory(conf)
+
+  override protected def estimate(info: FlowStrategyInfo, conf: JobConf, history: Seq[FlowStepHistory]): Option[MemoryEstimate] = {
+    // iterate over mem history
+    // collect: for maps, list of max memory in past runs
+    //          for reduce, list of max memory in past runs
+    // compute smoothed memory est
+    // multiple by scale factor
+    // cap estimate to max size if needed
+    // handle gc
+    // return
+    val alpha = MemoryEstimatorConfig.getAlpha(conf)
+    val scaleFactor = MemoryEstimatorConfig.getScaleFactor(conf)
+    val yarnSchedulerIncrementMB = MemoryEstimatorConfig.getYarnSchedulerIncrement(conf)
+
+    LOG.info(s"Attempting to estimate memory settings with alpha: $alpha, scaleFactor: $scaleFactor, schedulerIncrement: $yarnSchedulerIncrementMB")
+
+    val maxMemoryList: Seq[(Option[Long], Option[Long])] = history.map{ h => getMapReduceMemory(h) }
+    val maxMapList: Seq[Long] = maxMemoryList.flatMap(_._1)
+    val maxReduceList: Seq[Long] = maxMemoryList.flatMap(_._2)
+
+    val mapSmoothEst: Double = smoothedAverage(maxMapList, alpha)
+    val reduceSmoothEst: Double = smoothedAverage(maxReduceList, alpha)
```

Comment:
It might be interesting to expose something like these methods as protected methods, as I can imagine we may be interested in trying out different ideas for this:

```
trait MemoryHistoryEstimator extends HistoryEstimator[MemoryEstimate] {
  // ...
  def estimateMapMemory(history: Seq[(Time, Long)]): Option[Long]
  def estimateReducerMemory(history: Seq[(Time, Long)]): Option[Long]
}

class SmoothedMemoryHistoryEstimator(alpha: Double) extends MemoryHistoryEstimator {
  def estimateMapMemory(history: Seq[(Time, Long)]): Option[Long] = // ...
  def estimateReducerMemory(history: Seq[(Time, Long)]): Option[Long] = // ...
}
```